### PR TITLE
add npmrc and enforce node and npm versions

### PR DESCRIPTION
**Pull request by Jessica-Koch**
Enforce a node version above 16, lts is 16.14.  When a user goes to run a yarn install in any of the directories, an error will be returned: 
image

I would like input on whether I should add an upper constraint of 17.0, v16 goes into maintenance in october: 
image

### CODE REVIEW

**Comment by aeunms** [comment]
When you say upper constraint of 17, do you mean versions less than 17? Since odd versions are not stable

**Comment by TrevorJoelHarris** [comment]
Supporting 16 for now makes sense since it's the active one. As Angela pointed out, since 17 isn't stable I don't think we want to restrict to just 17.

Personally, I wouldn't block people using 17 if they want to unless we know it will cause a problem.

**Review by TrevorJoelHarris — APPROVED** [review]

**Comment by erinha** [comment]
I believe we have github actions that verify that our build works on node versions 12 and 14, which are now failing. Additionally, our pipeline runs on node 14, so that appears to be failing as well. Those will have to be updated before this change is merged. 

Note that the github actions that verify the build works on node 12 and 14 are the same as those running in v1, so we should double-check that there isn't a reason we should continue supporting 12 and 14.

File: .npmrc
Comment (erinha) [inline .npmrc:3]: What is the effect if we don't have `engine-strict=true`? Can we display warnings without failing the build?

**Review by erinha — CHANGES_REQUESTED** [review]
Appreciate the intent here to help us all stay consistent/avoid problems due to node versions! Question about `engine-strict=true` and comment about failing pipelines/checks.

File: package.json
Comment (erinha) [inline package.json:12]: This is probably a stupid question, but since we use yarn instead of npm, what is the reason to include this here?

**Comment by erinha** [comment]
> Supporting 16 for now makes sense since it's the active one. As Angela pointed out, since 17 isn't stable I don't think we want to restrict to just 17.
> 
> Personally, I wouldn't block people using 17 if they want to unless we know it will cause a problem.

Jess, were you asking if we should block people from using versions greater than 17. So that for example, when version 18 is released, if anyone upgraded too soon, it would cause failures and notify us that we need to update it? Or am I misunderstanding?

File: .npmrc
Comment (Jessica-Koch) [reply .npmrc:3]: It force displays the errors in the console, we need this in order to see them, the .npmrc file defines the settings on how NPM should behave when running commands.

File: package.json
Comment (Jessica-Koch) [reply package.json:12]: Not a stupid question, yarn is installed via npm, so this impacts yarn (I tested it because I was confused at first), the screenshot above is running a yarn command in the console.

**Comment by Jessica-Koch** [comment]
I can't respond to comments in this UI so I'll respond to them here

@aeunms - I didn't realize that the odd versions weren't stable so that is interesting!  I meant making it so a dev can only use a node version that is greater than 15 and less than 17.  Right now, I have it configured to only throw an error when the node version is less than 16. 

@TrevorJoelHarris - you answered my question, that's what I was asking, should we stop people from using 17, I know personally I was originally using 17 and it did throw errors, but maybe because it isn't stable, that should be expected.  I like the idea of allowing people to use the newer version, it makes updating later easier.  

@erinha - Yes I did notice that these pipelines are failing, you're totally right that these would have to be updated.  This brought up a question that I have which is, if we are testing the code against older versions of node, then why when a dev uses v14, for instance, do they get issues if the code will pass checks in our pipeline?  This may be obvious, I just don't understand.  I would love to have a conversation about the best way forward in terms of updating these pipelines and whatnot, who else I would need to talk to etc.

**Comment by Jessica-Koch** [comment]
> > Supporting 16 for now makes sense since it's the active one. As Angela pointed out, since 17 isn't stable I don't think we want to restrict to just 17.
> > Personally, I wouldn't block people using 17 if they want to unless we know it will cause a problem.
> 
> @Jessica-Koch, were you asking if we should block people from using versions greater than 17. So that for example, when version 18 is released, if anyone upgraded too soon, it would cause failures and notify us that we need to update it? Or am I misunderstanding?

I was asking if we should stop people from using versions 17 and above, but honestly, because those aren't stable maybe it should be assumed that it may break the build.  I'm hesitant to restrict newer versions because I know from experience, it's easy to just never update the version of node and then you run into a problem when you notice and your version is 4 versions out of date.  If that makes any sense at all.  If we keep the upper bound open, then we can more easily test with newer node versions when 17 becomes stable (or I guess becomes 18?  referring to what @aeunms  said above)

**Comment by erinha** [comment]
> why when a dev uses v14, for instance, do they get issues if the code will pass checks in our pipeline? 

@Jessica-Koch this is a great question. If they pass the pipeline I would expect them to pass in our code as well. Were people seeing local issues with node versions 12 and 14? I thought it was just unsupported versions.

**Comment by erinha** [comment]
> I would love to have a conversation about the best way forward in terms of updating these pipelines and whatnot, who else I would need to talk to etc.

To update the ADO pipelines, you just need to edit the yaml files that reference the node version. For removing the github action checks that validate the build on node versions 12 and 14, I would recommend getting approval/input from @ydogandjiev.

File: .npmrc
Comment (erinha) [reply .npmrc:3]: That's helpful, thanks. Do you know if it's possible to make it display warnings but not errors?

Comment (Jessica-Koch) [reply .npmrc:3]: I think I answered a different question than the one you asked, i'll look into this because i'm not sure

Comment (Jessica-Koch) [reply .npmrc:3]: My thought though is that adding the strict true is better because it's alerting a problem before it starts, if we make it a warning, it may end up breaking the build later and just kicks the problem down the road.  If this is a warning, i'm not sure I see the point of changing this at all...warnings get ignored (like the ones on the main page of this repo warning us about vulnerabilities that still have gone unchanged)

Comment (Jessica-Koch) [reply .npmrc:3]: I may not be looking at the benefits of having a warning correctly, but the way I'm seeing it, just making it a warning so it passes the build isn't the best justification, but I am totally open to being wrong about this

**Comment by Jessica-Koch** [comment]
I am closing this as I don't think it is useful and forgot to close this with the ticket

**Pull request closed without merging**
